File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex14DataFreshness.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-14 — Data freshness and report metadata.
/// </summary>
/// <remarks>
/// Source timestamps, report date and request timing, feeding an application-owned
/// re-review rule.
/// </remarks>
public sealed class Ex14DataFreshness : IExample
{
    /// <inheritdoc />
    public string Id => "ex-14";

    /// <inheritdoc />
    public string Title => "Source timestamps feeding an application-owned re-review rule.";

    /// <summary>How old this application lets a source get before it wants another look.</summary>
    private const int StaleAfterDays = 180;

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        foreach (var ein in new[] { Fixtures.Eins.PublicCharity, Fixtures.Eins.StaleData })
        {
            var result = await context.Client.Nonprofits.CheckAsync(ein);
            var organization = result.Nonprofit!;

            Output.Heading(Output.Text(organization.OrganizationName));

            var sources = new (string Label, string Field)[]
            {
                ("Publication 78 extract", "most_recent_pub78"),
                ("Business Master File extract", "most_recent_bmf"),
                ("Pactman record modified", "organization_info_last_modified"),
                ("Report compiled", "report_date"),
            };

            var stale = new List<string>();
            var unparseable = new List<string>();

            foreach (var (label, field) in sources)
            {
                var raw = organization.GetString(field);
                var age = ApiDate.AgeInDays(raw);

                Output.Field(
                    label,
                    (raw, age) switch
                    {
                        (null, _) when !organization.Has(field) => Output.NotReturned,
                        (null, _) => "null",
                        (_, null) => $"{raw} (could not be parsed)",
                        _ => $"{raw} — {age} days ago",
                    });

                if (raw is not null && age is null)
                {
                    // A date that will not parse is not a fresh date. Say so rather than
                    // treating it as absent and moving on.
                    unparseable.Add(field);
                }
                else if (age > StaleAfterDays)
                {
                    stale.Add(field);
                }
            }

            Output.Field("request timing", $"{result.TimeTakenMs} ms server-side");

            Output.Heading("The re-review rule");
            Output.Field("threshold", $"{StaleAfterDays} days");
            Output.Field("stale sources", stale.Count == 0 ? "none" : string.Join(", ", stale));
            Output.Field("unparseable", unparseable.Count == 0 ? "none" : string.Join(", ", unparseable));
            Output.Field(
                "routes to",
                stale.Count > 0 || unparseable.Count > 0 ? "re-review" : "accept");
        }

        Output.Note(
            "The API publishes when each source was extracted; how old is too old is your\n"
            + "policy, not the SDK's. Note that a fresh report_date over a stale extract\n"
            + "means the report is new and the underlying IRS data is not.");
    }
}

File: ruby/examples/ex_14_data_freshness.rb
# frozen_string_literal: true

# EX-14 — Data freshness and report metadata.
#
# Every source on the response carries its own date. A check is a statement about
# the data as of those dates — not as of the moment you called.
#
# This example surfaces each timestamp, computes an age, and applies a re-review
# rule the application owns. The SDK supplies the dates and nothing else: there
# is no `stale?` method and no default threshold, because 90 days is prudent for
# one workflow and reckless for another.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_14_data_freshness.rb

require "time"
require_relative "lib/api_date"
require_relative "lib/fixture_api"
require_relative "lib/print"

include Print

# The application's own rule. Change it here, in one place.
RE_REVIEW_AFTER_DAYS = 90

TIMESTAMP_FIELDS = %w[organization_info_last_modified report_date most_recent_bmf most_recent_pub78].freeze

def age_in_days(value, now)
  parsed = ApiDate.parse(value)
  parsed && ApiDate.age_in_days(parsed, now)
end

FixtureApi.with_fixture_api do |client|
  [
    ["recently refreshed", FixtureApi::EINS[:public_charity]],
    ["every source is old", FixtureApi::EINS[:stale_data]],
    ["some dates were not returned", FixtureApi::EINS[:sparse_identity]]
  ].each do |label, ein|
    result = client.nonprofits.check(ein)
    nonprofit = result.nonprofit

    if nonprofit.nil?
      puts "No record for #{ein}."
      next
    end

    now = Time.now
    timestamps = TIMESTAMP_FIELDS.to_h { |name| [name, nonprofit[name]] }

    heading "#{label} — #{nonprofit.organization_name}"

    timestamps.each do |name, value|
      age = age_in_days(value, now)

      puts "  #{name.ljust(34)} #{(value || "<null>").ljust(26)} #{age.nil? ? "age unknown" : "#{age} days old"}"
    end

    # `report_date` is when this response was generated. The source dates are
    # when each underlying list was last refreshed. They answer different
    # questions, and the older one governs.
    ages = timestamps.filter_map do |name, value|
      age = age_in_days(value, now)
      { name: name, age: age } unless age.nil?
    end

    oldest = ages.max_by { |entry| entry[:age] }
    undated = timestamps.select { |_name, value| value.nil? || value.empty? }.keys

    bullet "oldest source: #{oldest[:name]} at #{oldest[:age]} days" if oldest
    undated.each { |name| bullet "no date returned for #{name} — age cannot be established" }

    needs_re_review = (oldest.nil? || oldest[:age] > RE_REVIEW_AFTER_DAYS) || !undated.empty?

    field "request timing (timeTaken ms)", result.time_taken_ms
    field "checked at (UTC)", now.utc.iso8601
    field "re-review rule (> #{RE_REVIEW_AFTER_DAYS} days)",
          if needs_re_review
            "schedule a re-review — a source is past the threshold or undated"
          else
            "within the freshness window — no re-review scheduled"
          end

    # Store the timestamps alongside your verification record, not just the
    # outcome. Six months from now "we checked and it was fine" is not an answer;
    # "we checked on this date against BMF data published on that date" is.
    next unless ein == FixtureApi::EINS[:public_charity]

    puts "\n  stored with the verification record:"
    json_block({ "ein" => nonprofit.ein, "checked_at" => now.utc.iso8601, "request_id" => result.request_id }
                 .merge(timestamps))
  end
end

note "A fresh response is not a fresh fact. IRS lists publish on their own schedule,\n" \
     "so a check performed today can reflect a revocation posted weeks ago and not\n" \
     "yet published — see ex_29 for the pre-payment recheck this implies."
